shell/common/rasterizer.h

```diff
@@ -20,78 +20,367 @@
+  ///
+  //  TODO(chinmaygarde): The rasterizer does not use the task runners for
+  //  anything other than thread checks. Remove the same as an argument.
+  ///
+  /// @param[in]  delegate            The rasterizer delegate.
+  /// @param[in]  task_runners        The task runners used by the shell.
+  /// @param[in]  compositor_context  The compositor context used to hold all
+  ///                                 the GPU state used by the rasterizer.
+  ///
   Rasterizer(Delegate& delegate,
              TaskRunners task_runners,
              std::unique_ptr<flutter::CompositorContext> compositor_context);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Destroys the rasterizer. This must happen on the GPU task
+  ///             runner. All GPU resources are collected before this call
+  ///             returns. Any context setup by the embedder to hold these
+  ///             resources can be immediately collected as well.
+  ///
   ~Rasterizer();
 
+  //----------------------------------------------------------------------------
+  /// @brief      Rasterizers may be created well before an on-screen surface is
+  ///             available for rendering. Shells usually create a rasterizer in
+  ///             their constructors. Once an on-screen surface is available
+  ///             however, one may be provided to the rasterizer using this
+  ///             call. No rendering may occur before this call. The surface is
+  ///             held till the balancing call to `Rasterizer::Teardown` is
+  ///             made. Calling a setup before tearing down the previous surface
+  ///             (if this is not the first time the surface has been setup) is
+  ///             user error.
+  ///
+  /// @see        `Rasterizer::Teardown`
+  ///
+  /// @param[in]  surface  The on-screen render surface.
+  ///
   void Setup(std::unique_ptr<Surface> surface);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Releases the previously setup on-screen render surface and
+  ///             collects associated resources. No more rendering may occur
+  ///             till the next call to `Rasterizer::Setup` with a new render
+  ///             surface. Calling a teardown without a setup is user error.
+  ///
   void Teardown();
 
-  // Frees up Skia GPU resources.
-  //
-  // This method must be called from the GPU task runner.
+  //----------------------------------------------------------------------------
+  /// @brief      Notifies the rasterizer that there is a low memory situation
+  ///             and it must purge as many unnecessary resources as possible.
+  ///             Currently, the Skia context associated with onscreen rendering
+  ///             is told to free GPU resources.
+  ///
   void NotifyLowMemoryWarning() const;
 
+  //----------------------------------------------------------------------------
```

Comment:
Maybe add "The weak pointer can be copied to other threads, but the rasterizer it references to may only be accessed..."

Reply:
Isn't that a property of all weak pointers? That is documented in FML. This weak pointer is not special in any way.